sysvabi64/sysvabi64.rst

```diff
@@ -720,22 +724,37 @@ open-source compilers.
 
 .. table:: Code Model Support
 
-  +-------+----------------+----------------+
-  | Code  | GCC 11         | Clang 12.0     |
-  | Model |                |                |
-  +=======+================+================+
-  | tiny  | yes            | yes            |
-  |       |                |                |
-  +-------+----------------+----------------+
-  | small | pic and PIC    | pic and PIC    |
-  |       | implemented    | accepted but   |
-  |       |                | PIC used       |
-  +-------+----------------+----------------+
-  | large | no-pic support | no-pic support |
-  |       |                |                |
-  |       |                |                |
-  |       |                |                |
-  +-------+----------------+----------------+
+  +--------+----------------+----------------+
+  | Code   | GCC 11         | Clang 12.0     |
+  | Model  |                |                |
+  +========+================+================+
+  | tiny   | yes            | yes            |
+  |        |                |                |
+  +--------+----------------+----------------+
+  | small  | pic and PIC    | pic and PIC    |
+  |        | implemented    | accepted but   |
+  |        |                | PIC used       |
+  +--------+----------------+----------------+
+  | medium | pic and PIC    | pic and PIC    |
+  |        | proposed       | proposed       |
+  |        |                |                |
+  +--------+----------------+----------------+
+  | large  | no-pic support | no-pic support |
+  |        |                |                |
+  |        |                |                |
+  |        |                |                |
+  +--------+----------------+----------------+
+
+
+Medium code model
+-----------------
+
+Symbols smaller than 64KiB (value TBD) are treated like the small code model.
+Symbols which are larger than 64KiB, are ``common`` or of unknown size use
+sections ``.ldata``, ``.lbss`` and ``.lrodata``. These sections are placed after
+all other sections and can be unlimited in size. Symbols in them are addressed
+via the GOT. Linkers may relax GOT indirections into PC-relative addressing for
+non pre-emptible symbols within 4GiB.
 
```

Comment:
If I've understood correctly we'll end up with an image layout (simplified) like:
.text
.rodata
.rwdata
.bss
.lrodata
.lrwdata
.lbss

This would mean at least a couple more page-aligned segments. It would also mean linker's needing modifications in generic places (linker scripting) to recognise the sections, we'd probably need a more distinctive name to avoid name clahes. By default they'd be placed as orphans matched to the nearest type (.lrodata after .rodata) etc.  

Is that worth it? For example could this be simplified to everything being accessed via the GOT, with the GOT relaxation to the PC relative addressing mode doing the heavy lifting. This might mean introducing a new relocation directive to make the relaxation easier to do in more cases. For example R\_\<CLS\>\_ADR_GOTPAGERELAX and R\_\<CLS\>\_LD64_GOTRELAX\_LO12\_NC which a linker only needs to check if there is a non-pre-emptable definition of the symbol before deciding to relax?

I can see a case when there are a few really gigantic sections which it would be advisable to sort later. In terms of the requirements, perhaps it would be better to just say that sections should have a prefix of .ldata, .lbss and .lrodata and let the programmer write a linker script if they felt like sorting these after. This would make this proposal easier to deploy. The prefix is important as something like -fdata-sections will name sections .data.<variable-name>.

Reply:
Yes, in the medium model you end up with an extra load segment if you do have some large arrays. The linker could automatically merge large sections with the standard ones if they fit, and place them at the end if they don't. An alternative may be to place small symbols in sbss/sdata and ensure both come before data/bss - but that might also add an extra load segment?

Note lbss/ldata/sbss/sdata are supported on multiple targets, it is mostly turning on the sections and implementing target hooks. Since it's not new functionality I'd expect the implementation to be fairly straightforward.

Without a way to sort symbols into small and large, we would need to assume all symbols are out of range. That means we have an overhead on every symbol access rather than only on a few huge arrays (since it applies to static variables as well, it's even worse than -fPIC!). On a prototype sorting into different sections reduces the performance overhead by 3x and codesize overhead by 10x.

Yes, adding new relaxing relocations is certainly possible. However it means changes to compilers, assemblers and linkers to support new relocations, and swap between new/old relocations. This creates backwards compatibility issues which you avoid when relaxing existing relocations as an optional optimization. Note this should ideally be discussed on my other pull request since relaxation is not required for or exclusive to the medium model.

I'm not sure I get your point about writing a linker script - the medium model is meant to do that automatically for you.

Reply:
Thanks for pointing out that existing uses of .ldata etc. Looking at AArch64 and GNU ld I can see that X86_64 has .ldata etc. in its default linker script, but AArch64 does not, presumably part of the build-time configuration. LLD while not having a built-in linker script does not treat .ldata or .lbss separately for any architecture. That just means that we'll need to have linker work to implement this, or require a linker script if using an older linker. I note that the script puts the .lbss adjacent to the .bss which makes it more likely only a single additional segment is needed.

aarch64:
```
  . = SEGMENT_START("ldata-segment", .);
  . = ALIGN(64 / 8);
  __end__ = .;
  _end = .; PROVIDE (end = .);
  . = DATA_SEGMENT_END (.);
```

x86_64:
```
  .bss            :
  {
   *(.dynbss)
   *(.bss .bss.* .gnu.linkonce.b.*)
   *(COMMON)
   /* Align here to ensure that the .bss section occupies space up to
      _end.  Align after .bss to ensure correct alignment even if the
      .bss section disappears because there are no input sections.
      FIXME: Why do we need it? When there is no .bss section, we don't
      pad the .data section.  */
   . = ALIGN(. != 0 ? 64 / 8 : 1);
  }
  .lbss   :
  {
    *(.dynlbss)
    *(.lbss .lbss.* .gnu.linkonce.lb.*)
    *(LARGE_COMMON)
  }
  . = ALIGN(64 / 8);
  . = SEGMENT_START("ldata-segment", .);
  .lrodata   ALIGN(CONSTANT (MAXPAGESIZE)) + (. & (CONSTANT (MAXPAGESIZE) - 1)) :
  {
    *(.lrodata .lrodata.* .gnu.linkonce.lr.*)
  }
  .ldata   ALIGN(CONSTANT (MAXPAGESIZE)) + (. & (CONSTANT (MAXPAGESIZE) - 1)) :
  {
    *(.ldata .ldata.* .gnu.linkonce.l.*)
    . = ALIGN(. != 0 ? 64 / 8 : 1);
  }
  . = ALIGN(64 / 8);
  _end = .; PROVIDE (end = .);
  . = DATA_SEGMENT_END (.);`
```

Some thoughts:
* Should .ldata, .lbss etc be a mandatory part of the medium code-model or an optimisation of it? For example could it be QoI? For example implementations may choose to use ...

I'll add a comment to your relaxation proposal as well.

Reply:
If we set the threshold for .lrodata/.ldata large enough, you very rarely need them. Large arrays are typically uninitialized and thus use .lbss, so you might not need an extra segment in most cases.

The only potentially tricky part is handling of common symbols. Given common symbols are already placed at the end of .bss (as shown above), you could assume all common symbols are large and use a GOT indirection (since -fno-common is the default, only Fortan COMMON blocks will create common symbols).

I'm not sure how you could make .lbss optional since it results in binary incompatibility (eg. GCC and LLVM x86 medium models are incompatible).

Reply:
Just to check on binary compatibility? 

Do you mean that some programs may fail to link due to differing assumptions of placement. For example compiler 1 that uses GOT addressing everywhere but doesn't use .lbss when mixed with objects from compiler 2 that only uses GOT addressing for .lbss may end up with variables out of range due to large arrays from compiler 1 placed in .bss.

From looking at some descriptions of the x86_64 medium code model it does sound like it is only worth doing the medium code-model without .ldata and friends. If that is the case then fine; it is not worth trying to simplify it further.

Reply:
Indeed, these are incompatible ABIs so you can't mix and match them. This is why we have to clearly define a medium model for AArch64.

Which description do you mean? By using .ldata you only need a GOT indirection on large arrays, so almost all accesses are as efficient as the small code model (and since array addresses are typically lifted out of loops, the GOT indirection is fairly cheap). Without .ldata every global access would need a GOT indirection, so you have a similar overhead as PIC with -fPIE or -fno-pic. To optimize this you might need new relocations to avoid GOT indirections on every global access.

Reply:
The description I was referring to was https://eli.thegreenplace.net/2012/01/03/understanding-the-x64-code-models which has a section on Medium code model, at least the GCC variant of it.

Reply:
FWIW I'm happy with the proposal. I'll need to read through carefully to see if there are any suggestions for wording/claifications.

My understanding is that this code-model should do the majority of what the large code model does, but in a more efficient and position independent way.

Reply:
Note the medium model allows rodata of unlimited size which is not possible in the large model (text+rodata must stay <2GB due to eh_frame relocations - unless you swap eh_frame and ro_data placement if ro_data is large). So the new medium model effectively obsoletes the large model - there are no circumstances where you would prefer using the existing large model!